util/tracecpu/profile.go

```diff
@@ -0,0 +1,401 @@
+	} else {
+		ctx = pprof.WithLabels(context.Background(), pprof.Labels(labelSQLDigest, sqlDigest))
+	}
+	pprof.SetGoroutineLabels(ctx)
+	GlobalSQLStatsProfiler.RegisterSQL(sqlDigest, normalizedSQL)
+	return ctx
+}
+
+// SetGoroutineLabelsWithSQLAndPlan sets the SQL and plan digest label into the goroutine.
+func SetGoroutineLabelsWithSQLAndPlan(ctx context.Context, sqlDigest, planDigest, normalizedPlan string) context.Context {
+	ctx = pprof.WithLabels(ctx, pprof.Labels(labelSQLDigest, sqlDigest, labelPlanDigest, planDigest))
+	pprof.SetGoroutineLabels(ctx)
+	GlobalSQLStatsProfiler.RegisterPlan(planDigest, normalizedPlan)
+	return ctx
+}
+
+func (sp *sqlStatsProfiler) startExportCPUProfile(w io.Writer) error {
+	sp.mu.Lock()
+	defer sp.mu.Unlock()
+	if sp.mu.ept != nil {
+		return errors.New("cpu profiling already in use")
+	}
+	sp.mu.ept = &exportProfileTask{w: w}
+	return nil
+}
+
+func (sp *sqlStatsProfiler) stopExportCPUProfile() error {
+	sp.mu.Lock()
+	ept := sp.mu.ept
+	sp.mu.ept = nil
+	sp.mu.Unlock()
+	if ept.err != nil {
+		return ept.err
+	}
+	if w := ept.w; w != nil && ept.cpuProfile != nil {
+		sp.removeLabel(ept.cpuProfile)
+		return ept.cpuProfile.Write(w)
+	}
+	return nil
+}
+
+func (sp *sqlStatsProfiler) removeLabel(p *profile.Profile) {
+	if p == nil {
+		return
+	}
+	keepLabelSQL := variable.EnablePProfSQLCPU.Load()
+	for _, s := range p.Sample {
+		for k := range s.Label {
+			if keepLabelSQL && k == labelSQL {
+				continue
+			}
+			delete(s.Label, k)
+		}
+	}
+}
+
+type exportProfileTask struct {
+	cpuProfile *profile.Profile
+	err        error
+	w          io.Writer
+}
+
+func (t *exportProfileTask) mergeProfile(p *profile.Profile) {
+	if t.err != nil || p == nil {
+		return
+	}
+	ps := make([]*profile.Profile, 0, 2)
+	if t.cpuProfile != nil {
+		ps = append(ps, t.cpuProfile)
+	}
+	ps = append(ps, p)
+	t.cpuProfile, t.err = profile.Merge(ps)
+}
+
+// ProfileHTTPHandler is same as pprof.Profile.
+// The difference is ProfileHTTPHandler uses tracecpu.StartCPUProfile/StopCPUProfile to fetch profile data.
+func ProfileHTTPHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+	sec, err := strconv.ParseInt(r.FormValue("seconds"), 10, 64)
+	if sec <= 0 || err != nil {
+		sec = 30
+	}
+
+	if durationExceedsWriteTimeout(r, float64(sec)) {
+		serveError(w, http.StatusBadRequest, "profile duration exceeds server's WriteTimeout")
+		return
+	}
+
+	// Set Content Type assuming StartCPUProfile will work,
+	// because if it does it starts writing.
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Disposition", `attachment; filename="profile"`)
+
+	err = StartCPUProfile(w)
+	if err != nil {
+		serveError(w, http.StatusInternalServerError, "Could not enable CPU profiling: "+err.Error())
+		return
+	}
+	time.Sleep(time.Second * time.Duration(sec))
```

Comment:
The background profile collector runs periodically, so the profile sample obtained are less than expected?

Reply:
Not obtained less, but the profile timeline does not work as expected. Already add a comment and `TODO` to fix it in next PR.